audio/src/range_set.rs

```diff
@@ -1,124 +1,138 @@
+        for range in &self.ranges {
+            write!(f, "{}", range)?;
+        }
+        write!(f, ")")
     }
 }
 
-impl Range {
-    pub fn new(start: usize, length: usize) -> Range {
-        Range { start, length }
-    }
+/// `RangeSet` can implement `Index`, but not `IndexMut`, since that would allow the user to modify
+/// the ranges in the set, which may fail to uphold the invariant around ordering.
+impl std::ops::Index<usize> for RangeSet {
+    type Output = Range;
 
-    pub fn end(&self) -> usize {
-        self.start + self.length
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.ranges[index]
     }
 }
 
-#[derive(Clone)]
-pub struct RangeSet {
-    ranges: Vec<Range>,
+impl IntoIterator for RangeSet {
+    type Item = Range;
+
+    type IntoIter = <Vec<Range> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.ranges.into_iter()
+    }
 }
 
-impl fmt::Display for RangeSet {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "(").unwrap();
-        for range in self.ranges.iter() {
-            write!(f, "{}", range).unwrap();
+impl From<Vec<Range>> for RangeSet {
+    fn from(mut ranges: Vec<Range>) -> Self {
+        ranges.sort_unstable_by_key(|r| r.start());
+        let mut result: Vec<Range> = Vec::with_capacity(ranges.len()); // asssume worst case
+        for range in ranges {
+            match result.last_mut() {
+                Some(r) if r.touches(range) => *r = r.union(range),
+                _ => result.push(range),
+            }
         }
-        write!(f, ")")
+
+        Self { ranges: result }
     }
 }
 
 impl RangeSet {
+    /// Construct a new `RangeSet`. Equivalent to `RangeSet::default()`.
     pub fn new() -> RangeSet {
-        RangeSet {
-            ranges: Vec::<Range>::new(),
-        }
+        Self::default()
     }
 
+    /// Checks whether this `RangeSet` contains no values.
     pub fn is_empty(&self) -> bool {
         self.ranges.is_empty()
     }
 
+    /// Returns the total length of all ranges contained in this `RangeSet`.
     pub fn len(&self) -> usize {
-        self.ranges.iter().map(|r| r.length).sum()
+        self.ranges.iter().map(Range::len).sum()
     }
 
-    pub fn get_range(&self, index: usize) -> Range {
-        self.ranges[index]
-    }
-
-    pub fn iter(&self) -> Iter<Range> {
+    /// Returns a borrowed iterator over the ranges in this `RangeSet`.
+    pub fn iter(&self) -> impl Iterator<Item = &Range> {
         self.ranges.iter()
     }
 
+    /// Returns `true` if the value is contained, returns `false` otherwise.
     pub fn contains(&self, value: usize) -> bool {
-        for range in self.ranges.iter() {
-            if value < range.start {
-                return false;
-            } else if range.start <= value && value < range.end() {
-                return true;
-            }
-        }
-        false
+        self.iter().any(|r| r.contains(value))
```

Comment:
If you want to (it's really not necessary), take a look at https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search_by

Reply:
Yes I binary search would be faster since our data is ordered. Good idea! I have implemented this with a closure containing a match, let me know what you think.